### PR TITLE
Set an initial value index during join of MincLoc, MaxLoc or MinMaxLoc

### LINKED ISSUE #6144: {Min,Max,MinMax}Loc needs a value index at initialization
Consider a `MinLoc` reduction over `View<unsigned int*> v` that's entirely filled with 0, followed by an access to `v(reduction_result.location)`. The reducer/reduction implementation should guarantee that the index produced is valid (outside the degenerate case of `v.extent(0) == 0`). This means that the initialization of those reducers must use some valid index value, and not an out-of-bounds sentinel, because code that says `if (v(i) < result.value) { result.value = v(i); result.location = i; }` may never overwrite it.

### PR DESCRIPTION
**Pull request by ldh4**
Resolves #6144

This PR sets a value index during `join` of `MincLoc`, `MaxLoc` or `MinMaxLoc`, if the source values are always equivalent to the comparing destination value, in which case the index was never updated and remained either `reduction_identity<index_type>::min()` or `reduction_identity<index_type>::max()`.

Added unit tests for `MinLoc`, `MaxLoc` and `MinMaxLoc` based on a reproducer: https://github.com/kokkos/kokkos/issues/6144#issuecomment-2341915033.

Co-Authored-By: Christian Trott <crtrott@sandia.gov>

### CODE REVIEW

File: core/unit_test/TestReducers.hpp [hunk 1]
```diff
@@ -1396,6 +1513,7 @@ struct TestReducers {
 #endif
 #else
     test_minmaxloc(10007);
+    test_minmaxloc_loc_init(101);
```
[the remark this answers is not in the document]
Comment (ldh4) [reply core/unit_test/TestReducers.hpp]: Reduced the value of N to 3 and checked that the failure does happen without the changes in `join` for OpenMP build with 4 threads.

File: core/unit_test/TestReducers.hpp [hunk 2]
```diff
@@ -823,6 +823,37 @@ struct TestReducers {
...
+            return;
+          else if (x <= update.val) {
+            update.val = x;
+            update.loc = i;
+          }
+        },
+        reducer_type(value_loc));
+
+    ASSERT_EQ(value_loc.val, h_values(0));
+    ASSERT_TRUE(value_loc.loc >= 0 && value_loc.loc < N);
```
[the remark this answers is not in the document]
Comment (ldh4) [reply core/unit_test/TestReducers.hpp]: Applied.

**Review by masterleinad — COMMENTED** [review]
We should also test
- MinMaxFirstLastLoc
- MinFirstLoc
- MaxFirstLoc

**Comment by ldh4** [comment]
> We should also test
> 
> * MinMaxFirstLastLoc
> * MinFirstLoc
> * MaxFirstLoc

Ok, I'll add those for tests as well. 
It seems that these are not listed in [Kokkos built-in reducers](https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/Custom-Reductions-Built-In-Reducers.html?highlight=built+reducers). Are these not public?

**Comment by masterleinad** [comment]
> It seems that these are not listed in [Kokkos built-in reducers](https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/Custom-Reductions-Built-In-Reducers.html?highlight=built+reducers). Are these not public?

The should be public and we should add them to the documentation.

**Comment by ldh4** [comment]
> We should also test
> 
> * MinMaxFirstLastLoc
> * MinFirstLoc
> * MaxFirstLoc

Added equivalent tests for these built-in reducers.
With changes in this PR, `MinLoc` is now functionally equivalent to `MinFirstLoc`, and `MaxLoc` is equivalent to `MaxFirstLoc`.

**Comment by masterleinad** [comment]
> With changes in this PR, `MinLoc` is now functionally equivalent to `MinFirstLoc`, and `MaxLoc` is equivalent to `MaxFirstLoc`.

Is that really true? I thought `MinLoc` still gives you *some* valid index that has the minimum value while you get the *minimum* index that has the minimum value with `MinFirstLoc`. It look to me that we still don't enforce anything more than a valid index for the former.
It might be valid to ask why we need both, though ,(assuming there is not much of a difference in performance).

**Comment by ldh4** [comment]
> I thought `MinLoc` still gives you _some_ valid index that has the minimum value while you get the _minimum_ index that has the minimum value with `MinFirstLoc`. 

You are right. I overlooked the fact that `join` may not necessarily be called with serialized order of index of `Loc`. `MinLoc` will simply set its index to the first valid index it's called with.

**Review by crtrott — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
OpenMPTarget failed the new tests in a real way: @rgayatri23 can you please take a look at this?

**Comment by masterleinad** [comment]
> OpenMPTarget failed the new tests in a real way: @rgayatri23 can you please take a look at this?

```diff
diff --git a/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp b/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp
index 9b578aca1..f7623398a 100644
--- a/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp
+++ b/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp
@@ -236,12 +236,12 @@ struct OpenMPTargetReducerWrapper<MinLoc<Scalar, Index, Space>> {
   // Required
   KOKKOS_INLINE_FUNCTION
   static void join(value_type& dest, const value_type& src) {
-    if (src.val < dest.val) dest = src;
-  }
-
-  KOKKOS_INLINE_FUNCTION
-  static void join(volatile value_type& dest, const volatile value_type& src) {
-    if (src.val < dest.val) dest = src;
+      if (src.val < dest.val)
+      dest = src;
+    else if (src.val == dest.val &&
+             dest.loc == reduction_identity<index_type>::min()) {
+      dest.loc = src.loc;
+    }
   }
 
   KOKKOS_INLINE_FUNCTION
@@ -263,12 +263,12 @@ struct OpenMPTargetReducerWrapper<MaxLoc<Scalar, Index, Space>> {
 
   KOKKOS_INLINE_FUNCTION
   static void join(value_type& dest, const value_type& src) {
-    if (src.val > dest.val) dest = src;
-  }
-
-  KOKKOS_INLINE_FUNCTION
-  static void join(volatile value_type& dest, const volatile value_type& src) {
-    if (src.val > dest.val) dest = src;
+     if (src.val > dest.val)
+      dest = src;
+    else if (src.val == dest.val &&
+             dest.loc == reduction_identity<index_type>::min()) {
+      dest.loc = src.loc;
+    }
   }
 
   KOKKOS_INLINE_FUNCTION
@@ -328,25 +328,19 @@ struct OpenMPTargetReducerWrapper<MinMaxLoc<Scalar, Index, Space>> {
   // Required
   KOKKOS_INLINE_FUNCTION
   static void join(value_type& dest, const value_type& src) {
-    if (src.min_val < dest.min_val) {
+	      if (src.min_val < dest.min_val) {
       dest.min_val = src.min_val;
       dest.min_loc = src.min_loc;
-    }
-    if (src.max_val > dest.max_val) {
-      dest.max_val = src.max_val;
-      dest.max_loc = src.max_loc;
-    }
-  }
-
-  KOKKOS_INLINE_FUNCTION
-  static void join(volatile value_type& dest, const volatile value_type& src) {
-    if (src.min_val < dest.min_val) {
-      dest.min_val = src.min_val;
+    } else if (dest.min_val == src.min_val &&
+               dest.min_loc == reduction_identity<index_type>::min()) {
       dest.min_loc = src.min_loc;
     }
     if (src.max_val > dest.max_val) {
       dest.max_val = src.max_val;
       dest.max_loc = src.max_loc;
+    } else if (dest.max_val == src.max_val &&
+               dest.max_loc == reduction_identity<index_type>::min()) {
+      dest.max_loc = src.max_loc;
     }
   }
```
Would fix it indeed.

**Comment by rgayatri23** [comment]
> > OpenMPTarget failed the new tests in a real way: @rgayatri23 can you please take a look at this?
> 
> ```diff
> diff --git a/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp b/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp
> index 9b578aca1..f7623398a 100644
> --- a/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp
> +++ b/core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp
> @@ -236,12 +236,12 @@ struct OpenMPTargetReducerWrapper<MinLoc<Scalar, Index, Space>> {
>    // Required
>    KOKKOS_INLINE_FUNCTION
>    static void join(value_type& dest, const value_type& src) {
> -    if (src.val < dest.val) dest = src;
> -  }
> -
> -  KOKKOS_INLINE_FUNCTION
> -  static void join(volatile value_type& dest, const volatile value_type& src) {
> -    if (src.val < dest.val) dest = src;
> +      if (src.val < dest.val)
> +      dest = src;
> +    else if (src.val == dest.val &&
> +             dest.loc == reduction_identity<index_type>::min()) {
> +      dest.loc = src.loc;
> +    }
>    }
>  
>    KOKKOS_INLINE_FUNCTION
> @@ -263,12 +263,12 @@ struct OpenMPTargetReducerWrapper<MaxLoc<Scalar, Index, Space>> {
>  
>    KOKKOS_INLINE_FUNCTION
>    static void join(value_type& dest, const value_type& src) {
> -    if (src.val > dest.val) dest = src;
> -  }
> -
> -  KOKKOS_INLINE_FUNCTION
> -  static void join(volatile value_type& dest, const volatile value_type& src) {
> -    if (src.val > dest.val) dest = src;
> +     if (src.val > dest.val)
> +      dest = src;
> +    else if (src.val == dest.val &&
> +             dest.loc == reduction_identity<index_type>::min()) {
> +      dest.loc = src.loc;
> +    }
>    }
>  
>    KOKKOS_INLINE_FUNCTION
> @@ -328,25 +328,19 @@ struct OpenMPTargetReducerWrapper<MinMaxLoc<Scalar, Index, Space>> {
>    // Required
>    KOKKOS_INLINE_FUNCTION
>    static void join(value_type& dest, const value_type& src) {
> -    if (src.min_val < dest.min_val) {
> +	      if (src.min_val < dest.min_val) {
>        dest.min_val = src.min_val;
>        dest.min_loc = src.min_loc;
> -    }
> -    if (src.max_val > dest.max_val) {
> -      dest.max_val = src.max_val;
> -      dest.max_loc = src.max_loc;
> -    }
> -  }
> -
> -  KOKKOS_INLINE_FUNCTION
> -  static void join(volatile value_type& dest, const volatile value_type& src) {
> -    if (src.min_val < dest.min_val) {
> -      dest.min_val = src.min_val;
> +    } else if (dest.min_val == src.min_val &&
> +               dest.min_loc == reduction_identity<index_type>::min()) {
>        dest.min_loc = src.min_loc;
>      }
>      if (src.max_val > dest.max_val) {
>        dest.max_val = src.max_val;
>        dest.max_loc = src.max_loc;
> +    } else if (dest.max_val == src.max_val &&
> +               dest.max_loc == reduction_identity<index_type>::min()) {
> +      dest.max_loc = src.max_loc;
>      }
>    }
> ```
> 
> Would fix it indeed.

LGTM. This actually looks like a bug that should be fixed.

**Review by rgayatri23 — APPROVED** [review]

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp [hunk 3]
```diff
@@ -236,12 +236,12 @@ struct OpenMPTargetReducerWrapper<MinLoc<Scalar, Index, Space>> {
   // Required
   KOKKOS_INLINE_FUNCTION
   static void join(value_type& dest, const value_type& src) {
-    if (src.val < dest.val) dest = src;
-  }
-
-  KOKKOS_INLINE_FUNCTION
-  static void join(volatile value_type& dest, const volatile value_type& src) {
-    if (src.val < dest.val) dest = src;
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Reducer.hpp:244]: That's not realli necessary here but I would want to remove the `volatile` overloads anyway.

**Comment by masterleinad** [comment]
`OPENACC-NVHPC-CUDA` segfaulted in the compiler on `fetnat03`.

**Comment by ldh4** [comment]
For OpenACC build: `Error: no GPU available for execution.`

For NVCC-debug:
```
29: [ RUN      ] cuda.dyn_rank_view_api_generic
29: Constructor for Kokkos::View 'drv0' has mismatched number of arguments. The number of arguments = 8 neither matches the dynamic rank = 7 nor the total rank = 7
29: 
29: Backtrace:
29: [0x7ffff79c014a] Kokkos::Impl::save_stacktrace()
29: [0x7ffff796e25f] Kokkos::Impl::host_abort(char const*)
```
Possibly related to the ongoing fix with DynRankView refactor?

**Review by crtrott — APPROVED** [review]

**Pull request merged**